src/module.c

```diff
@@ -2486,6 +2506,342 @@ RedisModuleString *RM_ListPop(RedisModuleKey *key, int where) {
+  */
+int RM_ListIteratorStop(RedisModuleKey *key) {
+    if (!key) {
+        errno = EINVAL;
+        return REDISMODULE_ERR;
+    } else if (!key->value || key->value->type != OBJ_LIST) {
+        errno = ENOTSUP;
+        return REDISMODULE_ERR;
+    } else if (!key->iter) {
+        errno = EBADF;
+        return REDISMODULE_ERR;
+    }
+    listTypeReleaseIterator(key->iter);
+    key->iter = NULL;
+    moduleDelKeyIfEmpty(key);
+    return REDISMODULE_OK;
+}
+
+/* Advances a list iterator and fetches the next element.
+ *
+ * On success, REDISMODULE_OK is returned and the `elem` pointer is set to the
+ * next element. The element should be free'd using RM_FreeString() or using
+ * automatic memory management. To advance the iterator without receiving the
+ * element, set `elem` to NULL. On failure, REDISMODULE_ERR is returned and
+ * `errno` is set accordningly:
+ *
+ * - EINVAL if key is NULL.
+ * - ENOTSUP if the key is not a list.
+ * - EBADF if no iterator is started.
+ * - ENOENT if there are no more elements in the list.
+ */
+int RM_ListIteratorNext(RedisModuleKey *key, RedisModuleString **elem) {
+    if (!key) {
+        errno = EINVAL;
+        return REDISMODULE_ERR;
+    } else if (!key->value || key->value->type != OBJ_LIST) {
+        errno = ENOTSUP;
+        return REDISMODULE_ERR;
+    } else if (!(key->mode & REDISMODULE_READ) || !key->iter) {
+        errno = EBADF;
+        return REDISMODULE_ERR;
+    }
+
+    listTypeIterator *li = key->iter;
+    if (key->u.list.reseek) {
+        /* listTypeInsert() has made the iterator invalid. Fix it. */
+        listTypeReseekIterator(li, key->u.list.index);
+        key->u.list.reseek = 0;
+    }
+    key->u.list.index += li->direction == LIST_HEAD ? -1 : 1;
+    if (listTypeNext(li, &key->u.list.entry)) {
+        key->u.list.hascurrent = 1;
+        if (elem != NULL) {
+            robj *encoded = listTypeGet(&key->u.list.entry);
+            *elem = getDecodedObject(encoded);
+            decrRefCount(encoded);
+            serverAssert((*elem)->type == OBJ_STRING);
+            autoMemoryAdd(key->ctx, REDISMODULE_AM_STRING, *elem);
+        }
+        return REDISMODULE_OK;
+    } else {
+        key->u.list.hascurrent = 0;
+        errno = ENOENT;
+        return REDISMODULE_ERR;
+    }
+}
+
+/* Deletes the list element at the current position in a list iterator started
+ * on a key.
+ *
+ * Returns REDISMODULE_OK on success. Returns REDISMODULE_ERR on failure and
+ * sets `errno` as follows:
+ *
+ * - EINVAL if key is NULL.
+ * - ENOTSUP if the key is not a list.
+ * - EBADF is no iterator has been started.
+ * - ENOENT if there is no current element.
+ */
+int RM_ListIteratorDelete(RedisModuleKey *key) {
+    if (!key) {
+        errno = EINVAL;
+        return REDISMODULE_ERR;
+    } else if (!key->value || key->value->type != OBJ_LIST) {
+        errno = ENOTSUP;
+        return REDISMODULE_ERR;
+    } else if (!(key->mode & REDISMODULE_READ) || !key->iter) {
```

Comment:
Check REDISMODULE_WRITE? Maybe with a designated EACCESS errno?

Reply:
Good catch.

I also considered EROFS (read-only filesystem) but I settled with EBADF when I implemented the Streams API (already merged).

EBADF is set by `write()` when "fd is not a valid file descriptor or is not open for writing". An opened key is like a file descriptor.

Are you OK with that?

Reply:
Yes, makes sense